File: prototype/rails/src/test/domain/escrow/fee_test.rb
require "test_helper"

module Domain
  module Escrow
    class FeeTest < ActiveSupport::TestCase
      test "the platform takes a tenth" do
        assert_equal 4500, Fee.platform(Money.from_cents(45_000)).cents
      end

      test "the seller keeps the rest" do
        assert_equal 40_500, Fee.net(Money.from_cents(45_000)).cents
      end

      test "the fee and the net add back up" do
        subtotal = Money.from_cents(4999)

        assert_equal subtotal.cents, Fee.platform(subtotal).cents + Fee.net(subtotal).cents
      end

      test "half a cent rounds away from zero" do
        assert_equal 5, Fee.platform(Money.from_cents(45)).cents
      end

      test "nothing owes nothing" do
        assert_equal 0, Fee.platform(Money.from_cents(0)).cents
      end
    end
  end
end

File: prototype/node/src/app/core/escrow/fee.test.ts
import { test } from 'node:test'
import assert from 'node:assert/strict'
import { platformFee, sellerNet } from './fee.ts'
import { cents } from '../money.ts'

test('platformFee takes a tenth', () => {
  assert.equal(platformFee(cents(45_000)), 4500)
})

test('sellerNet keeps the rest', () => {
  assert.equal(sellerNet(cents(45_000)), 40_500)
})

test('the fee and the net add back up to the subtotal', () => {
  const subtotal = cents(4999)

  assert.equal(platformFee(subtotal) + sellerNet(subtotal), subtotal)
})

test('half a cent rounds away from zero', () => {
  assert.equal(platformFee(cents(45)), 5)
})

test('nothing owes nothing', () => {
  assert.equal(platformFee(cents(0)), 0)
})

test('a negative subtotal takes a negative fee', () => {
  assert.equal(platformFee(cents(-45_000)), -4500)
})

test('sellerNet keeps the rest of a negative subtotal', () => {
  assert.equal(sellerNet(cents(-45_000)), -40_500)
})
